crates/router_env/src/request_id.rs

```diff
@@ -0,0 +1,365 @@
+        /// The invalid header value that caused the error.
+        value: String,
+    },
+}
+
+impl error_stack::Context for Error {}
+
+impl ResponseError for Error {}
+
+impl Display for Error {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::NoAssociatedId => write!(fmt, "No request ID associated with this request"),
+            Self::InvalidHeaderValue { value } => write!(fmt, "Invalid header value: {}", value),
+        }
+    }
+}
+
+/// Configuration for handling incoming request ID headers.
+#[derive(Clone, Copy, PartialEq, Eq, Default, Debug, serde::Deserialize, serde::Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum IdReuse {
+    /// Reuse the incoming request ID if present, otherwise generate a new one.
+    UseIncoming,
+    /// Always generate a new request ID, ignoring any incoming headers.
+    #[default]
+    IgnoreIncoming,
+}
+
+/// Generate a new UUID v7 request ID.
+fn generate_uuid_v7() -> String {
+    Uuid::now_v7().to_string()
+}
+
+/// Request ID middleware that takes a configurable header name
+/// and determines how to handle incoming request IDs.
+#[derive(Clone, Debug)]
+pub struct RequestIdentifier {
+    header_name: String,
+    use_incoming_id: IdReuse,
+}
+
+/// Request ID value that can be extracted in route handlers.
+///
+/// Wraps an `Arc<str>` for optimal performance in web middleware.
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub struct RequestId(Arc<str>);
+
+/// The middleware implementation that processes requests.
+#[derive(Debug)]
+pub struct RequestIdMiddleware<S> {
+    service: S,
+    header_name: HeaderName,
+    use_incoming_id: IdReuse,
+}
+
+impl Display for RequestId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.as_str())
+    }
+}
+
+impl FromStr for RequestId {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.is_empty() {
+            Err(Error::InvalidHeaderValue {
+                value: s.to_string(),
+            })
+        } else {
+            Ok(Self(s.into()))
+        }
+    }
+}
+
+impl TryFrom<HeaderValue> for RequestId {
+    type Error = Error;
+
+    fn try_from(value: HeaderValue) -> Result<Self, Self::Error> {
+        let s = value.to_str().map_err(|_| Error::InvalidHeaderValue {
+            value: format!("{:?}", value),
+        })?;
+        Self::from_str(s)
+    }
+}
+
+impl From<RequestId> for HeaderValue {
+    fn from(request_id: RequestId) -> Self {
+        Self::from_str(&request_id.0).unwrap_or_else(|_| Self::from_static("invalid-request-id"))
+    }
+}
+
+impl From<RequestId> for String {
+    fn from(request_id: RequestId) -> Self {
+        request_id.0.to_string()
+    }
+}
+
+impl From<String> for RequestId {
+    fn from(s: String) -> Self {
+        Self(s.into())
+    }
+}
+
+impl From<&str> for RequestId {
+    fn from(s: &str) -> Self {
+        Self(s.into())
+    }
+}
```

Comment:
We can't have these implementations as `From` implementations, right? These would have to be `TryFrom` implementations? To handle the empty string case at least?